woocommerce/woocommerce
# Add a string encoding selector to the product importer

### All Submissions:


### Changes proposed in this Pull Request:

The product importer tries to guess the string encoding of the supplied CSV file and falls back to UTF-8 when it's unable to do so. This is unreliable since it's not really possible to success in the detection for encodings like ISO-8859-1 (for European languages) which are incompatible with UTF-8, thus the strings containing non-ASCII characters in these files will be imported incorrectly.

This pull request adds a new dropdown to the first step of the product importer, under "advanced options", that allows to manually select the character encoding of the input file from the list of encodings supported by the server. Default is "Autodetect" which just relies on autodetection, which was the previous behavior.

Inspired by: https://github.com/woocommerce/woocommerce/pull/36416 (which can't be accepted because it uses [`utf8_encode`](https://www.php.net/manual/en/function.utf8-encode.php), which always uses ISO-8859-1 and is deprecated).

![image](https://user-images.githubusercontent.com/937723/218515527-b186f335-c1fd-42e7-98f4-015462e2f82e.png)

![image](https://user-images.githubusercontent.com/937723/218516029-77245839-e8db-400c-a5eb-7f61fee565af.png)


### How to test the changes in this Pull Request:

1. Create a product dump file that has non-ASCII character both in the row headers and in the product names, save it using the ISO-8859-1 encoding. Example file contents, supplied in a comment in https://github.com/woocommerce/woocommerce/pull/36416:

```
ID,Nom,Stock,Tarif régulier
555,Pét' Nat' Pépin Batch 4 - Domaine Achillée *,123
```

2. Go to the products page, click the "Import" button on the top of the page.

3. Notice the new "Character encoding of the file" dropdown, leave it as "Autodetect".

4. Import the file you just created (you need to manually map the "Nom" column to "Name")

  * Notice how in the "Column mapping" screen there's no sample value for the "Nom" (name) column, and no name for the last column ("Tarif régulier")

  * Notice how the name of the product has "?" character instead of "é" in the products page.

5. Repeat the import but this time select ISO-8859-1 (or Windows-1252) in the encoding selection dropdown.

  * Notice how now the product name appears as a sample for the name column in "Column mapping" and the last column has the proper "Tarif régulier" name in the column mappings screen.

  * Notice how the name of the product is now correct in the products page.

6. Repeat but this time converting the file to UTF-8 and leaving the encoding as "Autodetect", UTF-8 was already supported and it should still work.


### Other information:


### FOR PR REVIEWER ONLY:
